File: internal/installer/steps/tailscale.go
package steps

import (
	"context"
	"fmt"
	"runtime"

	"github.com/SolsticeSauer/solsclaw_beta/internal/installer"
)

type Tailscale struct{}

func (Tailscale) ID() string    { return "tailscale" }
func (Tailscale) Label() string { return "Install Tailscale + register node" }

func (Tailscale) ShouldRun(sc installer.StepContext) bool {
	// Docker mode handles Tailscale via a sidecar container; no host
	// install needed.
	if sc.Submission.InstallMode == installer.ModeDocker {
		return false
	}
	return sc.Submission.OptionalFeatures.Tailscale.Enabled
}

func (s Tailscale) Run(ctx context.Context, sc installer.StepContext) error {
	if !commandExists("tailscale") {
		if err := installTailscale(ctx, sc, s.ID()); err != nil {
			return err
		}
	} else {
		sc.Bus.Log(s.ID(), "info", "Tailscale already installed; skipping install.")
	}

	args := []string{"up", "--accept-routes"}
	ts := sc.Submission.OptionalFeatures.Tailscale
	if ts.AuthKey != "" {
		args = append(args, "--authkey="+ts.AuthKey)
	}
	if ts.Hostname != "" {
		args = append(args, "--hostname="+ts.Hostname)
	}
	if err := streamCommand(ctx, sc, s.ID(), "tailscale", args...); err != nil {
		return err
	}

	// Auth keys are sensitive — drop the in-memory copy now that we're done.
	sc.Submission.OptionalFeatures.Tailscale.AuthKey = ""
	return nil
}

func installTailscale(ctx context.Context, sc installer.StepContext, stepID string) error {
	switch runtime.GOOS {
	case "darwin":
		sc.Bus.Log(stepID, "info", "Installing Tailscale via Homebrew...")
		return streamCommand(ctx, sc, stepID, "brew", "install", "--cask", "tailscale")
	case "linux":
		sc.Bus.Log(stepID, "info", "Downloading https://tailscale.com/install.sh ...")
		staged, err := stage("https://tailscale.com/install.sh")
		if err != nil {
			return err
		}
		sc.Bus.Log(stepID, "info", fmt.Sprintf("Staged at %s (sha256=%s)", staged.Path, staged.SHA256))
		return streamCommand(ctx, sc, stepID, "sh", staged.Path)
	case "windows":
		sc.Bus.Log(stepID, "info", "Installing Tailscale via winget...")
		return streamCommand(ctx, sc, stepID, "winget", "install", "--silent", "--id", "tailscale.tailscale")
	}
	return fmt.Errorf("Tailscale install not supported on %s", runtime.GOOS)
}

File: packages/installer-server/src/lib/steps/tailscale.ts
import { execa } from 'execa';
import { stageScript } from '../downloader.js';
import type { InstallerStep } from './index.js';

async function isInstalled(): Promise<boolean> {
  try {
    await execa('tailscale', ['version'], { stdio: 'ignore' });
    return true;
  } catch {
    return false;
  }
}

async function installTailscale(
  platform: NodeJS.Platform,
  log: (level: 'info' | 'warn', msg: string) => void,
): Promise<void> {
  if (platform === 'darwin') {
    log('info', 'Installing Tailscale via Homebrew...');
    await execa('brew', ['install', '--cask', 'tailscale'], { stdio: 'inherit' });
    return;
  }
  if (platform === 'linux') {
    log('info', 'Downloading https://tailscale.com/install.sh...');
    const staged = await stageScript('https://tailscale.com/install.sh');
    log('info', `Staged at ${staged.path} (sha256=${staged.sha256}).`);
    await execa('sh', [staged.path], { stdio: 'inherit' });
    return;
  }
  if (platform === 'win32') {
    log('info', 'Installing Tailscale via winget...');
    await execa('winget', ['install', '--silent', '--id', 'tailscale.tailscale'], { stdio: 'inherit' });
    return;
  }
  throw new Error(`Tailscale install not supported on ${platform}`);
}

export const tailscaleStep: InstallerStep = {
  id: 'tailscale',
  label: 'Install Tailscale + register node',
  shouldRun: ({ submission }) => submission.optionalFeatures.tailscale.enabled,
  async run({ submission, bus, platform }) {
    if (!(await isInstalled())) {
      await installTailscale(platform, (level, m) => bus.log(tailscaleStep.id, level, m));
    } else {
      bus.log(tailscaleStep.id, 'info', 'Tailscale already installed, skipping install.');
    }

    const ts = submission.optionalFeatures.tailscale;
    const args = ['up', '--accept-routes'];
    if (ts.authKey) args.push(`--authkey=${ts.authKey}`);
    if (ts.hostname) args.push(`--hostname=${ts.hostname}`);

    bus.log(tailscaleStep.id, 'info', 'Bringing Tailscale up...');
    await execa('tailscale', args, { stdio: 'inherit' });

    // Auth keys are sensitive — drop reference once consumed.
    submission.optionalFeatures.tailscale.authKey = undefined;
  },
};
